src/sources/futures.rs

```diff
@@ -56,43 +116,187 @@ impl<T> Scheduler<T> {
+            //
+            //   The runnable is created on the origin thread and sent to the origin thread, since both
+            //   Scheduler and Executor are !Send and !Sync. The waker may be sent to another thread,
+            //   which means that the scheduler function (and the Runnable it handles) can exist on
+            //   another thread. However, the scheduler function immediately sends it back to the origin
+            //   thread.
+            //
+            //   The issue then becomes "is the Runnable dropped unsoundly when the channel is closed?"
+            //   This problem is circumvented by immediately waking all runnables (pushes them into the
+            //   queue) and then draining the queue. This means that, before the channel is closed, all
+            //   runnables will be destroyed, preventing any unsoundness.
+            //
+            // - "If future is not 'static, borrowed variables must outlive its Runnable."
+            //
+            //   `F` is `'static`, so we don't have to worry about this one.
+            //
+            // - "If schedule is not Send and Sync, the task’s Waker must be used and dropped on the
+            //    original thread."
+            //
+            //   The schedule function uses a thread-safe MPSC channel to send the runnable back to the
+            //   origin thread. This means that the waker can be sent to another thread, which satisfies
+            //   this requirement.
+            //
+            // - "If schedule is not 'static, borrowed variables must outlive the task’s Waker."
+            //
+            //   `schedule` and the types it handles are `'static`, so we don't have to worry about
+            //   this one.
+            unsafe { builder.spawn_unchecked(move |_| future, schedule) }
+        };
+
+        // Insert the runnable into the set of active tasks.
+        active.insert(Active::Future(runnable.waker()));
+        drop(active_guard);
+
+        // Schedule the runnable and detach the task so it isn't cancellable.
+        runnable.schedule();
+        task.detach();
+
+        Ok(())
     }
 }
 
-/// Error generated when trying to schedule a future after the
-/// executor was destroyed.
-#[derive(thiserror::Error, Debug)]
-#[error("the executor was destroyed")]
-pub struct ExecutorDestroyed;
+impl Sender {
+    /// Send a runnable to the executor.
+    fn send(&self, runnable: Runnable<usize>) {
+        // Send on the channel.
+        if let Err(e) = self
+            .sender
+            .lock()
+            .unwrap_or_else(|e| e.into_inner())
+            .send(runnable)
+        {
+            // Make sure the runnable is never dropped.
+            std::mem::forget(e);
 
-#[derive(Debug)]
-struct ExecWaker {
-    ping: Ping,
+            // This shouldn't be able to happen since all of the tasks are destroyed before the
+            // executor is. This indicates a critical soundness bug.
+            std::process::abort();
```

Comment:
Is there a reason to `abort()` instead of panicking?

Reply:
This is a safe guard in the event that something I've overlooked causes unsound behavior. In this case, the Sender shouldn't fail, but if it does, it means that something has gone terribly unsound. In this case aborting is the only option, as panicking may let the program continue with unsound behavior.

Reply:
I see, but in that case we should at least print *something* to the user, that's why I believe panicking would be preferable: it gives relevant information to debug & understand what occurred.

Indeed the user may technically catch the panic and continue running, but it is widely admitted in the community that it's the responsibility of the one catching the panic to ensure that what they are doing is still safe. So I'd prefer a panic with a relevant error message: that sounds like a place for using `unreachable!()`.

Reply:
Sounds good; I panic instead.